crates/anonymization/src/core/hash.rs

```diff
@@ -0,0 +1,100 @@
+use argon2::Argon2;
+use base64::{engine::general_purpose, Engine as _};
+use sha2::{Digest, Sha256};
+use tiny_keccak::{Hasher as _, Sha3};
+
+use crate::{ano_error, core::AnoError};
+
+// Available hashing methods
+#[derive(PartialEq, Eq)]
+pub enum HashMethod {
+    /// Represents the SHA2 hash method with an optional salt.
+    SHA2(Option<Vec<u8>>),
+    /// Represents the SHA3 hash method with an optional salt.
+    SHA3(Option<Vec<u8>>),
+    /// Represents the Argon2 hash method with a mandatory salt.
+    Argon2(Vec<u8>),
+}
+
+impl HashMethod {
+    /// `HashMethod` constructor for interfaces
+    ///
+    /// * `method` - The hash method to use. This can be one of the following:
+    ///   SHA2, SHA3 or Argon2
+    /// * `salt` - An optional salt to use. Required with Argon2
+    pub fn new(hasher_method: &str, salt_opt: Option<Vec<u8>>) -> Result<Self, AnoError> {
+        match hasher_method {
+            "SHA2" => Ok(Self::SHA2(salt_opt)),
+            "SHA3" => Ok(Self::SHA3(salt_opt)),
+            "Argon2" => salt_opt.map_or_else(
+                || Err(ano_error!("Argon2 requires a salt value.")),
+                |salt| Ok(Self::Argon2(salt)),
+            ),
+            _ => Err(ano_error!("Not a valid hash method specified.")),
+        }
+    }
+}
+
+pub struct Hasher {
+    method: HashMethod, // The selected hash method
+}
+
+impl Hasher {
+    /// Creates a new `Hasher` instance using the specified hash method and an
+    /// optional salt.
+    ///
+    /// # Arguments
+    ///
+    /// * `method` - The hash method to use. This can be one of the following:
+    ///   * `SHA2` Fast and secure, but vulnerable to brute-force attacks.
+    ///   * `SHA3` Secure and resistant to brute-force attacks, but slower than
+    ///     SHA-256 and not as widely supported.
+    ///   * `Argon2` Highly resistant to brute-force attacks, but can be slower
+    ///     than other hash functions and may require more memory.
+    #[must_use]
+    pub const fn new(method: HashMethod) -> Self {
+        Self { method }
+    }
+
+    /// Applies the chosen hash method to the input data
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - A byte slice representing the input data to be hashed.
+    ///
+    /// # Returns
+    ///
+    /// The base64-encoded hash string.
+    pub fn apply(&self, data: &[u8]) -> Result<String, AnoError> {
```

Comment:
Good point but as seen with Manu, the goal of this library is not to make a generic hash function (otherwise on could use the hash methods already available in Rust) but to expose functions to anonymize data and output it as characters writable in a CSV file.

Reply:
I understand, but imho the string conversion should be performed in the writer function (separation of concerns -> what if you want to write a CSV with a different encoding?). Your code still has value since you create a uniform interface to several hashing methods, and manage the salt.

Reply:
I replaced the apply function with `apply_bytes` and `apply_str`